### ISSUE TITLE
Testing AppImages without libfuse2 installed in the host, but only exported in ~/.bashrc

### ISSUE THREAD

**Issue by ivan-hc** [ivan-hc/AM#124]
@probonopd I've performed a test today, just because I was curious

First of all I have uninstalled libfuse2 on my host system, so my Appimages were not working.

Then I have performed the following commands to download a .deb package of libfuse2, I have extracted it and placed the "libfuse2*.so*" libraries into a dedicated directory in my $HOME named "libfuse2" (being me lack of fantasy) and then I have exported such directory  in my ~/.bashrc file (you can do it manually, being that directory named libfuse2 the only one on my system, I've found the path using "find"):
```
mkdir dl-fuse2-tmp
cd dl-fuse2-tmp
LIBFUSE2_DEB=$(wget -q http://ftp.debian.org/debian/pool/main/f/fuse/ -O - | grep -Po '(?<=href=")[^"]*' | sort | grep -v exp | grep amd64 | grep "libfuse2_" | tail -1)
wget http://ftp.debian.org/debian/pool/main/f/fuse/"$LIBFUSE2_DEB"
ar x ./*.deb
tar fx ./data.tar.xz
cd ..
mkdir libfuse2
mv ./dl-fuse2-tmp/lib/*/libfuse* ./libfuse2/
rm -R -f ./dl-fuse2-tmp
LIBFUSE2_DIR_PATH=$(find / -type d -name libfuse2 2>/dev/null | head -1)
echo -e '\nexport  LD_LIBRARY_PATH='$LIBFUSE2_DIR_PATH':${LD_LIBRARY_PATH}\n' >> $HOME/.bashrc
```
Now my AppImages are working.

I use Debian Testing, the downloaded package come from Debian Sid and this is the output related to GLIBC:
```
ivan@debian:~/libfuse2$ ldd -v libfuse.so.2 | grep GLIBC
		libc.so.6 (GLIBC_2.28) => /lib/x86_64-linux-gnu/libc.so.6
		libc.so.6 (GLIBC_2.32) => /lib/x86_64-linux-gnu/libc.so.6
		libc.so.6 (GLIBC_2.7) => /lib/x86_64-linux-gnu/libc.so.6
		libc.so.6 (GLIBC_2.14) => /lib/x86_64-linux-gnu/libc.so.6
		libc.so.6 (GLIBC_2.4) => /lib/x86_64-linux-gnu/libc.so.6
		libc.so.6 (GLIBC_2.33) => /lib/x86_64-linux-gnu/libc.so.6
		libc.so.6 (GLIBC_2.17) => /lib/x86_64-linux-gnu/libc.so.6
		libc.so.6 (GLIBC_2.3.2) => /lib/x86_64-linux-gnu/libc.so.6
		libc.so.6 (GLIBC_2.5) => /lib/x86_64-linux-gnu/libc.so.6
		libc.so.6 (GLIBC_2.34) => /lib/x86_64-linux-gnu/libc.so.6
		libc.so.6 (GLIBC_2.3.4) => /lib/x86_64-linux-gnu/libc.so.6
		libc.so.6 (GLIBC_2.2.5) => /lib/x86_64-linux-gnu/libc.so.6
		ld-linux-x86-64.so.2 (GLIBC_2.2.5) => /lib64/ld-linux-x86-64.so.2
		ld-linux-x86-64.so.2 (GLIBC_2.3) => /lib64/ld-linux-x86-64.so.2
		ld-linux-x86-64.so.2 (GLIBC_PRIVATE) => /lib64/ld-linux-x86-64.so.2
```
It seems that this approach made it possibile to use Appimages without having to install libraries at system level but only locally.

Maybe someone can mantain that version by adding security patches, and maybe it can be forked by someone to do something different just to made it work with Appimages.

In my use case, I can add an installation script in my AppImage managers AM and AppMan to made all these apps work without libfuse2 installed in /usr/lib

I'm not sure if all I'm sayng is correct. Someone else should test this on other distros.

**Comment by ivan-hc** [comment]
In my tests, after reboot, my installed AppImages can work if launched from terminal normally (without `--appimage-extract-and-run`), not from launchers.

**Comment by ivan-hc** [comment]
@zen0bit @DLopezJr @nazdridoy I invite you here to ask for confirmation, just follow the steps in the first comment, you can copy/paste the script I wrote above, it will create a directory named libfuse2 in the directory you will perfor the test (for example in your $HOME) and a string will be added in the ~/.bashrc

First of all, you have to remove libfuse2 with your package manager.

I've tested this on Debian Testing, if you have something different (Fedora, Arch, Slackware...) it would be useful.

**Comment by zen0bit** [comment]
Will test on void linux later today..

**Comment by ivan-hc** [comment]
is the environment variable in your ~/.bashrc?

Also you must close/open the terminal to see if it works.

**Comment by ivan-hc** [comment]
In the meantime I've found that each library must be loaded in /etc/ld.so.conf.d/libc.conf and then `sudo ldconfig` to update the library cache.

So root privileges are needed for this.

I've created and tested a script on my PC that downloads and extracts libfuse2 in /user/local/lib and updates the cache, so you have not to close the terminal:

https://github.com/ivan-hc/AM/assets/88724353/81792081-dea8-4fa2-ad95-72b677e71651

-----------------------------------------

This is the content:
```
#!/usr/bin/env bash

LIB=libfuse2

# CREATE THE FOLDER
mkdir /opt/$LIB
cd /opt/$LIB

# ADD THE REMOVER
echo '#!/bin/sh
rm -R -f /opt/'$LIB' /usr/local/lib/libfuse*' >> /opt/$LIB/remove
chmod a+x /opt/$LIB/remove

# DOWNLOAD THE ARCHIVE
mkdir tmp
cd ./tmp

LIBFUSE2_DEB=$(wget -q http://ftp.debian.org/debian/pool/main/f/fuse/ -O - | grep -Po '(?<=href=")[^"]*' | sort | grep -v exp | grep amd64 | grep "libfuse2_" | tail -1)
wget http://ftp.debian.org/debian/pool/main/f/fuse/"$LIBFUSE2_DEB"
ar x ./*.deb
tar fx ./data.tar.xz
cd ..
mkdir -p /usr/local/lib
mv ./tmp/lib/*/libfuse* /usr/local/lib/
rm -R -f ./tmp
ldconfig
```

it can be removed using the command `am -R libfuse2`

**Comment by ivan-hc** [comment]
> in debian its probably packaged with fuse

in Debian there is a dedicated package, `libfuse2`

From what you just said, it seems that libfuse2 is still an integrated part of the Arch Linux ecosystem, so if you have not fusermount you can't use any FUSE related library.

I have not `fusermount` in my repository, but I have `fuse3` and `libfuse3` installed. Those marked in green are all FUSE-related packages in my system (from Synaptics):

![Istantanea_2024-02-09_12-22-52](https://github.com/ivan-hc/AM/assets/88724353/5c66c378-70c3-41ad-806f-d97d05606d80)

**Comment by ivan-hc** [comment]
Have you seen if the library is in /usr/local/lib?

Extracting .deb packages requires `ar`, from `binutils`

**Comment by ivan-hc** [comment]
yes
```
ivan@debian:~$ find /usr/bin -name *fuse*
/usr/bin/fusermount3
/usr/bin/fusermount
/usr/bin/fuse-overlayfs
/usr/bin/fuser
```

**Comment by ivan-hc** [comment]
However, if you can't test an external libfuse2 library, there's not much we can do.

So Arch Linux is fine for now... maybe it should be tested on the ultimate anti-AppImages distribution: Fedora

Ubuntu is based on Debian, so the test should pass.

**Comment by ivan-hc** [comment]
I've tried running Appimage in a container in the past, without success. Docker, Podman, Distrobox, JuNest... nothing. Maybe I'm the one doing something wrong.

#### However, I don't want to turn this discussion into a tutorial for Distrobox or about package structure in the Arch and Debian repositories. Let's stay focused on the problem: we need to try to use libfuse2 as an external library and not a system one.

**Comment by zen0bit** [comment]
If you have good internet and space on drive you can easily try it yourself in any OS. Thanks to [quickemu](https://github.com/quickemu-project/quickemu) VM downloader and creator using qemu/kvm.
Project I contribute to, and also creating some TUIs for it

If you don't mind some other project promotion here.
But could be very useful for testing in different OSes

Will help you if you struggle using which I doubt :wink

**Comment by zen0bit** [comment]
I can't remove fuse.so.2 due to grub in void.
Which is default booting, so on base install can't be easily removed without also removing whole grub (booting) (could be used different one, but thats another story)

**Comment by ivan-hc** [comment]
> I can't remove fuse.so.2 due to grub in void.
Which is default booting, so on base install can't be easily removed without also removing whole grub (could be used different one, but thats another story)

I had the same issue on Debian Stable (32-bits) on Virtualbox

These tests came to mind since Testing is no longer an essential package in Debian. It's always present in the repositories, but I don't know for how long. Since I have been using Debian (almost 15 years) I have seen several programs considered obsolete disappear.

**Comment by zen0bit** [comment]
I am something like three years on linux exclusively

**Comment by ivan-hc** [comment]
For anyone want to test this...
### Requirements
 - `binutils`
 - `tar`
### Install
```
wget https://raw.githubusercontent.com/ivan-hc/AM/main/tools/libfuse2.sh
chmod a+x ./libfuse2.sh
sudo ./libfuse2.sh
```
### Uninstall
to unistall it, without AM
```
sudo /opt/libfuse2/remove
```
with AM
```
am -R libfuse2
```

in future I can upload installation scripts like this if someone needs a specific library not available on its distro's repository.

To work properly, you will need root (`sudo`) privileges, regardless of whether you use AM or AppMan

**Comment by ivan-hc** [comment]
I've performed a test using the library from `fuse2` in ArchLinux... and works on Debian:
```
function _debian_dl(){
	LIBFUSE2_DEB=$(wget -q http://ftp.debian.org/debian/pool/main/f/fuse/ -O - | grep -Po '(?<=href=")[^"]*' | sort | grep -v exp | grep amd64 | grep "libfuse2_" | tail -1)
	wget http://ftp.debian.org/debian/pool/main/f/fuse/"$LIBFUSE2_DEB"
	ar x ./*.deb
	tar fx ./data.tar.xz
	cd ..
	mkdir -p /usr/local/lib
	mv ./tmp/lib/*/libfuse* /usr/local/lib/
}

function _archlinux_dl(){
	ARCH=$(uname -m)
	wget https://archlinux.org/packages/extra/${ARCH}/fuse2/download && mv ./download ./$LIB.tar.zst
	tar fx ./*.tar.zst
	cd ..
	mkdir -p /usr/local/lib
	mv ./tmp/usr/lib/libfuse*2* /usr/local/lib/
}

#_debian_dl

_archlinux_dl
```

**Comment by ivan-hc** [comment]
Released 5.8 https://github.com/ivan-hc/AM/releases/tag/5.8
